File: native/src/option_chain.cpp
// option_chain.cpp - Option chain implementation
#include "option_chain.h"
#include "market_hours.h"
#include <algorithm>
#include <cmath>
#include <iterator>
#include <numeric>
#include <ql/instruments/vanillaoption.hpp>
#include <ql/pricingengines/blackformula.hpp>
#include <ql/pricingengines/vanilla/analyticeuropeanengine.hpp>
#include <ql/processes/blackscholesprocess.hpp>
#include <ql/quantlib.hpp>
#include <ql/termstructures/volatility/equityfx/blackconstantvol.hpp>
#include <ql/termstructures/yield/flatforward.hpp>
#include <spdlog/spdlog.h>

using namespace QuantLib;

// NOTE FOR AUTOMATION AGENTS:
// Option chain utilities live here. They transform raw market data into
// structures
// (`OptionChain`, `ExpiryChain`, `StrikeChain`) consumed by the strategy layer.
// Many quantitative routines remain stubbed; comments mark where
// production-grade Greeks/IV/DTE calculations should be wired in. Keep
// data-shaping logic centralized in this module so trading components can stay
// agnostic to quote representation.

namespace option_chain {

// ============================================================================
// OptionChainEntry Implementation
// ============================================================================

bool OptionChainEntry::is_valid() const {
  return contract.is_valid() && market_data.bid > 0 && market_data.ask > 0 &&
         market_data.ask >= market_data.bid;
}

bool OptionChainEntry::meets_liquidity_requirements(int min_volume,
                                                    int min_oi) const {
  return volume >= min_volume && open_interest >= min_oi;
}

// ============================================================================
// StrikeChain Implementation
// ============================================================================

double StrikeChain::get_call_iv() const {
  if (call.has_value() && call->market_data.implied_volatility.has_value()) {
    return call->market_data.implied_volatility.value();
  }
  return 0.0;
}

double StrikeChain::get_put_iv() const {
  if (put.has_value() && put->market_data.implied_volatility.has_value()) {
    return put->market_data.implied_volatility.value();
  }
  return 0.0;
}

double StrikeChain::get_iv_skew() const { return get_put_iv() - get_call_iv(); }

// ============================================================================
// ExpiryChain Implementation
// ============================================================================

ExpiryChain::ExpiryChain(const std::string &symbol, const std::string &expiry)
    : symbol_(symbol), expiry_(expiry) {}

void ExpiryChain::add_option(const OptionChainEntry &entry) {
  double strike = entry.contract.strike;

  auto &strike_chain = strikes_[strike];
  strike_chain.strike = strike;

  if (entry.contract.type == types::OptionType::Call) {
    strike_chain.call = entry;
  } else {
    strike_chain.put = entry;
  }
}

int ExpiryChain::get_days_to_expiry() const {
  if (expiry_.empty()) {
    return 0;
  }

  // Parse YYYYMMDD expiry string
  if (expiry_.length() != 8) {
    spdlog::warn("Invalid expiry format: {} (expected YYYYMMDD)", expiry_);
    return 0;
  }

  try {
    int year = std::stoi(expiry_.substr(0, 4));
    int month = std::stoi(expiry_.substr(4, 2));
    int day = std::stoi(expiry_.substr(6, 2));

    // Create expiry date
    std::tm tm_expiry = {};
    tm_expiry.tm_year = year - 1900;
    tm_expiry.tm_mon = month - 1;
    tm_expiry.tm_mday = day;
    tm_expiry.tm_hour = 16; // Market close (4:00 PM ET)
    tm_expiry.tm_min = 0;
    tm_expiry.tm_sec = 0;

    auto expiry_time =
        std::chrono::system_clock::from_time_t(std::mktime(&tm_expiry));
    auto now = std::chrono::system_clock::now();

    // Count trading days using MarketHours
    market_hours::MarketHours market_hours;
    int trading_days = 0;
    auto current = now;

    // If expiry is in the past, return 0
    if (expiry_time <= now) {
      return 0;
    }

    // Count trading days (weekdays excluding holidays)
    while (current < expiry_time) {
      auto status = market_hours.get_market_status_at(current);
      // Trading day = weekday and not holiday
      if (status.current_session != market_hours::MarketSession::Closed ||
          (!status.is_holiday && status.reason != "weekend")) {
        trading_days++;
      }
      current += std::chrono::hours(24);

      // Safety limit to prevent infinite loops
      if (trading_days > 365) {
        spdlog::warn("DTE calculation exceeded 365 days for expiry: {}",
                     expiry_);
        break;
      }
    }

    return trading_days;
  } catch (const std::exception &e) {
    spdlog::warn("Failed to calculate DTE for expiry {}: {}", expiry_,
                 e.what());
    return 0;
  }
}

std::vector<double> ExpiryChain::get_strikes() const {
  std::vector<double> strikes;
  for (const auto &[strike, _] : strikes_) {
    strikes.push_back(strike);
  }
  std::sort(strikes.begin(), strikes.end());
  return strikes;
}

std::optional<StrikeChain> ExpiryChain::get_strike_chain(double strike) const {
  auto it = strikes_.find(strike);
  if (it != strikes_.end()) {
    return it->second;
  }
  return std::nullopt;
}

std::optional<OptionChainEntry>
ExpiryChain::get_option(double strike, types::OptionType type) const {

  auto it = strikes_.find(strike);
  if (it != strikes_.end()) {
    if (type == types::OptionType::Call) {
      return it->second.call;
    } else {
      return it->second.put;
    }
  }
  return std::nullopt;
}

std::vector<double> ExpiryChain::get_strikes_in_range(double min_strike,
                                                      double max_strike) const {

  std::vector<double> strikes;
  for (const auto &[strike, _] : strikes_) {
    if (strike >= min_strike && strike <= max_strike) {
      strikes.push_back(strike);
    }
  }
  std::sort(strikes.begin(), strikes.end());
  return strikes;
}

std::optional<double>
ExpiryChain::find_atm_strike(double underlying_price) const {
  if (strikes_.empty()) {
    return std::nullopt;
  }

  double closest_strike = 0;
  double min_diff = std::numeric_limits<double>::max();

  for (const auto &[strike, _] : strikes_) {
    double diff = std::abs(strike - underlying_price);
    if (diff < min_diff) {
      min_diff = diff;
      closest_strike = strike;
    }
  }

  return closest_strike;
}

std::vector<OptionChainEntry> ExpiryChain::get_all_options() const {
  std::vector<OptionChainEntry> options;
  for (const auto &[_, chain] : strikes_) {
    if (chain.call.has_value()) {
      options.push_back(chain.call.value());
    }
    if (chain.put.has_value()) {
      options.push_back(chain.put.value());
    }
  }
  return options;
}

std::vector<OptionChainEntry> ExpiryChain::get_calls() const {
  std::vector<OptionChainEntry> calls;
  for (const auto &[_, chain] : strikes_) {
    if (chain.call.has_value()) {
      calls.push_back(chain.call.value());
    }
  }
  return calls;
}

std::vector<OptionChainEntry> ExpiryChain::get_puts() const {
  std::vector<OptionChainEntry> puts;
  for (const auto &[_, chain] : strikes_) {
    if (chain.put.has_value()) {
      puts.push_back(chain.put.value());
    }
  }
  return puts;
}

std::vector<OptionChainEntry>
ExpiryChain::filter_by_liquidity(int min_volume, int min_open_interest) const {

  auto options = get_all_options();
  std::vector<OptionChainEntry> filtered;
  filtered.reserve(options.size());

  std::copy_if(options.begin(), options.end(), std::back_inserter(filtered),
               [min_volume, min_open_interest](const OptionChainEntry &option) {
                 return option.meets_liquidity_requirements(min_volume,
                                                            min_open_interest);
               });

  return filtered;
}

std::vector<OptionChainEntry>
ExpiryChain::filter_by_moneyness(double min_moneyness,
                                 double max_moneyness) const {

  auto options = get_all_options();
  std::vector<OptionChainEntry> filtered;
  filtered.reserve(options.size());

  std::copy_if(options.begin(), options.end(), std::back_inserter(filtered),
               [min_moneyness, max_moneyness](const OptionChainEntry &option) {
                 return option.moneyness >= min_moneyness &&
                        option.moneyness <= max_moneyness;
               });

  return filtered;
}

// ============================================================================
// OptionChain Implementation
// ============================================================================

OptionChain::OptionChain(const std::string &symbol) : symbol_(symbol) {}

void OptionChain::add_option(const OptionChainEntry &entry) {
  auto it = expiries_.find(entry.contract.expiry);
  if (it == expiries_.end()) {
    auto [iter, inserted] = expiries_.emplace(
        entry.contract.expiry, ExpiryChain(symbol_, entry.contract.expiry));
    iter->second.add_option(entry);
  } else {
    it->second.add_option(entry);
  }
}

std::vector<std::string> OptionChain::get_expiries() const {
  std::vector<std::string> expiries;
  for (const auto &[expiry, _] : expiries_) {
    expiries.push_back(expiry);
  }
  std::sort(expiries.begin(), expiries.end());
  return expiries;
}

std::optional<ExpiryChain>
OptionChain::get_expiry_chain(const std::string &expiry) const {

  auto it = expiries_.find(expiry);
  if (it != expiries_.end()) {
    return it->second;
  }
  return std::nullopt;
}

std::vector<std::string>
OptionChain::get_expiries_in_dte_range(int min_dte, int max_dte) const {

  std::vector<std::string> filtered;
  for (const auto &[expiry, chain] : expiries_) {
    int dte = chain.get_days_to_expiry();
    if (dte >= min_dte && dte <= max_dte) {
      filtered.push_back(expiry);
    }
  }
  std::sort(filtered.begin(), filtered.end());
  return filtered;
}

std::vector<OptionChainEntry> OptionChain::get_all_options() const {
  std::vector<OptionChainEntry> all_options;
  for (const auto &[_, chain] : expiries_) {
    auto options = chain.get_all_options();
    all_options.insert(all_options.end(), options.begin(), options.end());
  }
  return all_options;
}

void OptionChain::set_underlying_price(double price) {
  underlying_price_ = price;
}

int OptionChain::get_total_option_count() const {
  return static_cast<int>(get_all_options().size());
}

int OptionChain::get_expiry_count() const {
  return static_cast<int>(expiries_.size());
}

// ============================================================================
// OptionChainBuilder Implementation (Stubs for Black-Scholes)
// ============================================================================

OptionChain OptionChainBuilder::build_from_market_data(
    const std::string &symbol,
    const std::vector<types::OptionContract> &contracts,
    const std::map<std::string, types::MarketData> &market_data) {

  OptionChain chain(symbol);

  spdlog::debug(
      "Building option chain for {} ({} contracts, {} with market data)",
      symbol, contracts.size(), market_data.size());

  // Build chain entries from contracts and market data
  for (const auto &contract : contracts) {
    std::string contract_key = contract.to_string();
    auto market_data_it = market_data.find(contract_key);

    if (market_data_it == market_data.end()) {
      spdlog::trace("No market data for contract: {}", contract_key);
      continue;
    }

    OptionChainEntry entry;
    entry.contract = contract;
    entry.market_data = market_data_it->second;

    // Set default liquidity scores (can be enhanced with actual volume/OI data)
    entry.volume = 0;
    entry.open_interest = 0;
    entry.liquidity_score = 50.0; // Default score

    chain.add_option(entry);
  }

  spdlog::debug("Built option chain with {} expiries, {} total options",
                chain.get_expiry_count(), chain.get_total_option_count());

  return chain;
}

void OptionChainBuilder::calculate_metrics(OptionChainEntry &entry,
                                           double underlying_price,
                                           double risk_free_rate) {

  entry.moneyness = entry.contract.strike / underlying_price;

  const double mid = entry.market_data.get_mid_price();
  if (entry.contract.type == types::OptionType::Call) {
    entry.intrinsic_value = std::max(underlying_price - entry.contract.strike, 0.0);
  } else {
    entry.intrinsic_value = std::max(entry.contract.strike - underlying_price, 0.0);
  }
  entry.extrinsic_value = std::max(mid - entry.intrinsic_value, 0.0);
}

std::optional<double> OptionChainBuilder::calculate_implied_volatility(
    double option_price, double underlying_price, double strike,
    double time_to_expiry, double risk_free_rate,
    types::OptionType option_type) {

  if (time_to_expiry <= 0.0 || option_price <= 0.0 || underlying_price <= 0.0 ||
      strike <= 0.0) {
    return std::nullopt;
  }

  try {
    // Convert time_to_expiry (years) to QuantLib time
    Time maturity = time_to_expiry;

    // Calculate implied volatility using QuantLib's black formula
    Option::Type ql_type =
        (option_type == types::OptionType::Call) ? Option::Call : Option::Put;

    // Use blackFormulaImpliedStdDev for implied volatility
    // This calculates the implied standard deviation (volatility * sqrt(time))
    double implied_std_dev =
        blackFormulaImpliedStdDev(ql_type, strike, underlying_price,
                                  option_price, risk_free_rate, maturity);

    // Convert to annualized volatility
    double implied_vol = implied_std_dev / std::sqrt(maturity);

    // Validate result
    if (implied_vol > 0.0 &&
        implied_vol < 10.0) { // Reasonable bounds (0-1000%)
      return implied_vol;
    }

    return std::nullopt;
  } catch (const std::exception &e) {
    spdlog::warn("QuantLib implied volatility calculation failed: {}",
                 e.what());
    return std::nullopt;
  }
}

double OptionChainBuilder::calculate_theoretical_price(
    double underlying_price, double strike, double time_to_expiry,
    double volatility, double risk_free_rate, types::OptionType option_type) {

  if (time_to_expiry <= 0.0 || underlying_price <= 0.0 || strike <= 0.0 ||
      volatility < 0.0) {
    return 0.0;
  }

  try {
    // Set up QuantLib date and calendar
    Date today = Date::todaysDate();
    Calendar calendar = TARGET();
    DayCounter dayCounter = Actual365Fixed();

    // Calculate expiration date from time_to_expiry (years)
    int days_to_expiry = static_cast<int>(time_to_expiry * 365.0);
    Date expiration = today + days_to_expiry;

    // Create option type
    Option::Type ql_type =
        (option_type == types::OptionType::Call) ? Option::Call : Option::Put;

    // Create handles for spot, rate, dividend (0), and volatility
    Handle<Quote> spotHandle(
        ext::shared_ptr<Quote>(new SimpleQuote(underlying_price)));
    Handle<YieldTermStructure> rateHandle(ext::shared_ptr<YieldTermStructure>(
        new FlatForward(today, risk_free_rate, dayCounter)));
    Handle<YieldTermStructure> divHandle(ext::shared_ptr<YieldTermStructure>(
        new FlatForward(today, 0.0, dayCounter)));
    Handle<BlackVolTermStructure> volHandle(
        ext::shared_ptr<BlackVolTermStructure>(
            new BlackConstantVol(today, calendar, volatility, dayCounter)));

    // Create Black-Scholes-Merton process (spot, dividend, risk-free, vol)
    ext::shared_ptr<BlackScholesMertonProcess> process(
        new BlackScholesMertonProcess(spotHandle, divHandle, rateHandle,
                                      volHandle));

    // Create pricing engine
    ext::shared_ptr<PricingEngine> engine(new AnalyticEuropeanEngine(process));

    // Create payoff and exercise
    ext::shared_ptr<StrikedTypePayoff> payoff(
        new PlainVanillaPayoff(ql_type, strike));
    ext::shared_ptr<Exercise> exercise(new EuropeanExercise(expiration));

    // Create option and set pricing engine
    VanillaOption option(payoff, exercise);
    option.setPricingEngine(engine);

    // Calculate theoretical price
    double theoretical_price = option.NPV();

    return theoretical_price;
  } catch (const std::exception &e) {
    spdlog::warn("QuantLib theoretical price calculation failed: {}", e.what());
    return 0.0;
  }
}

double OptionChainBuilder::calculate_delta(double underlying_price,
                                           double strike, double time_to_expiry,
                                           double volatility,
                                           double risk_free_rate,
                                           types::OptionType option_type) {

  if (time_to_expiry <= 0.0 || underlying_price <= 0.0 || strike <= 0.0 ||
      volatility < 0.0) {
    return 0.0;
  }

  try {
    // Set up QuantLib date and calendar
    Date today = Date::todaysDate();
    Calendar calendar = TARGET();
    DayCounter dayCounter = Actual365Fixed();

    // Calculate expiration date
    int days_to_expiry = static_cast<int>(time_to_expiry * 365.0);
    Date expiration = today + days_to_expiry;

    // Create option type
    Option::Type ql_type =
        (option_type == types::OptionType::Call) ? Option::Call : Option::Put;

    // Create handles
    Handle<Quote> spotHandle(
        ext::shared_ptr<Quote>(new SimpleQuote(underlying_price)));
    Handle<YieldTermStructure> rateHandle(ext::shared_ptr<YieldTermStructure>(
        new FlatForward(today, risk_free_rate, dayCounter)));
    Handle<YieldTermStructure> divHandle(ext::shared_ptr<YieldTermStructure>(
        new FlatForward(today, 0.0, dayCounter)));
    Handle<BlackVolTermStructure> volHandle(
        ext::shared_ptr<BlackVolTermStructure>(
            new BlackConstantVol(today, calendar, volatility, dayCounter)));

    // Create process and engine
    ext::shared_ptr<BlackScholesMertonProcess> process(
        new BlackScholesMertonProcess(spotHandle, divHandle, rateHandle,
                                      volHandle));
    ext::shared_ptr<PricingEngine> engine(new AnalyticEuropeanEngine(process));

    // Create option
    ext::shared_ptr<StrikedTypePayoff> payoff(
        new PlainVanillaPayoff(ql_type, strike));
    ext::shared_ptr<Exercise> exercise(new EuropeanExercise(expiration));
    VanillaOption option(payoff, exercise);
    option.setPricingEngine(engine);

    // Calculate delta
    return option.delta();
  } catch (const std::exception &e) {
    spdlog::warn("QuantLib delta calculation failed: {}", e.what());
    return 0.0;
  }
}

double OptionChainBuilder::calculate_gamma(double underlying_price,
                                           double strike, double time_to_expiry,
                                           double volatility,
                                           double risk_free_rate) {

  if (time_to_expiry <= 0.0 || underlying_price <= 0.0 || strike <= 0.0 ||
      volatility < 0.0) {
    return 0.0;
  }

  try {
    // Set up QuantLib (gamma is same for calls and puts)
    Date today = Date::todaysDate();
    Calendar calendar = TARGET();
    DayCounter dayCounter = Actual365Fixed();

    int days_to_expiry = static_cast<int>(time_to_expiry * 365.0);
    Date expiration = today + days_to_expiry;

    // Create handles
    Handle<Quote> spotHandle(
        ext::shared_ptr<Quote>(new SimpleQuote(underlying_price)));
    Handle<YieldTermStructure> rateHandle(ext::shared_ptr<YieldTermStructure>(
        new FlatForward(today, risk_free_rate, dayCounter)));
    Handle<YieldTermStructure> divHandle(ext::shared_ptr<YieldTermStructure>(
        new FlatForward(today, 0.0, dayCounter)));
    Handle<BlackVolTermStructure> volHandle(
        ext::shared_ptr<BlackVolTermStructure>(
            new BlackConstantVol(today, calendar, volatility, dayCounter)));

    // Create process and engine
    ext::shared_ptr<BlackScholesMertonProcess> process(
        new BlackScholesMertonProcess(spotHandle, divHandle, rateHandle,
                                      volHandle));
    ext::shared_ptr<PricingEngine> engine(new AnalyticEuropeanEngine(process));

    // Create option (use Call for gamma - same for both)
    ext::shared_ptr<StrikedTypePayoff> payoff(
        new PlainVanillaPayoff(Option::Call, strike));
    ext::shared_ptr<Exercise> exercise(new EuropeanExercise(expiration));
    VanillaOption option(payoff, exercise);
    option.setPricingEngine(engine);

    // Calculate gamma
    return option.gamma();
  } catch (const std::exception &e) {
    spdlog::warn("QuantLib gamma calculation failed: {}", e.what());
    return 0.0;
  }
}

double OptionChainBuilder::calculate_theta(double underlying_price,
                                           double strike, double time_to_expiry,
                                           double volatility,
                                           double risk_free_rate,
                                           types::OptionType option_type) {

  if (time_to_expiry <= 0.0 || underlying_price <= 0.0 || strike <= 0.0 ||
      volatility < 0.0) {
    return 0.0;
  }

  try {
    // Set up QuantLib
    Date today = Date::todaysDate();
    Calendar calendar = TARGET();
    DayCounter dayCounter = Actual365Fixed();

    int days_to_expiry = static_cast<int>(time_to_expiry * 365.0);
    Date expiration = today + days_to_expiry;

    Option::Type ql_type =
        (option_type == types::OptionType::Call) ? Option::Call : Option::Put;

    // Create handles
    Handle<Quote> spotHandle(
        ext::shared_ptr<Quote>(new SimpleQuote(underlying_price)));
    Handle<YieldTermStructure> rateHandle(ext::shared_ptr<YieldTermStructure>(
        new FlatForward(today, risk_free_rate, dayCounter)));
    Handle<YieldTermStructure> divHandle(ext::shared_ptr<YieldTermStructure>(
        new FlatForward(today, 0.0, dayCounter)));
    Handle<BlackVolTermStructure> volHandle(
        ext::shared_ptr<BlackVolTermStructure>(
            new BlackConstantVol(today, calendar, volatility, dayCounter)));

    // Create process and engine
    ext::shared_ptr<BlackScholesMertonProcess> process(
        new BlackScholesMertonProcess(spotHandle, divHandle, rateHandle,
                                      volHandle));
    ext::shared_ptr<PricingEngine> engine(new AnalyticEuropeanEngine(process));

    // Create option
    ext::shared_ptr<StrikedTypePayoff> payoff(
        new PlainVanillaPayoff(ql_type, strike));
    ext::shared_ptr<Exercise> exercise(new EuropeanExercise(expiration));
    VanillaOption option(payoff, exercise);
    option.setPricingEngine(engine);

    // Calculate theta (per day)
    return option.thetaPerDay();
  } catch (const std::exception &e) {
    spdlog::warn("QuantLib theta calculation failed: {}", e.what());
    return 0.0;
  }
}

double OptionChainBuilder::calculate_vega(double underlying_price,
                                          double strike, double time_to_expiry,
                                          double volatility,
                                          double risk_free_rate) {

  if (time_to_expiry <= 0.0 || underlying_price <= 0.0 || strike <= 0.0 ||
      volatility < 0.0) {
    return 0.0;
  }

  try {
    // Set up QuantLib (vega is same for calls and puts)
    Date today = Date::todaysDate();
    Calendar calendar = TARGET();
    DayCounter dayCounter = Actual365Fixed();

    int days_to_expiry = static_cast<int>(time_to_expiry * 365.0);
    Date expiration = today + days_to_expiry;

    // Create handles
    Handle<Quote> spotHandle(
        ext::shared_ptr<Quote>(new SimpleQuote(underlying_price)));
    Handle<YieldTermStructure> rateHandle(ext::shared_ptr<YieldTermStructure>(
        new FlatForward(today, risk_free_rate, dayCounter)));
    Handle<YieldTermStructure> divHandle(ext::shared_ptr<YieldTermStructure>(
        new FlatForward(today, 0.0, dayCounter)));
    Handle<BlackVolTermStructure> volHandle(
        ext::shared_ptr<BlackVolTermStructure>(
            new BlackConstantVol(today, calendar, volatility, dayCounter)));

    // Create process and engine
    ext::shared_ptr<BlackScholesMertonProcess> process(
        new BlackScholesMertonProcess(spotHandle, divHandle, rateHandle,
                                      volHandle));
    ext::shared_ptr<PricingEngine> engine(new AnalyticEuropeanEngine(process));

    // Create option (use Call for vega - same for both)
    ext::shared_ptr<StrikedTypePayoff> payoff(
        new PlainVanillaPayoff(Option::Call, strike));
    ext::shared_ptr<Exercise> exercise(new EuropeanExercise(expiration));
    VanillaOption option(payoff, exercise);
    option.setPricingEngine(engine);

    // Calculate vega
    return option.vega();
  } catch (const std::exception &e) {
    spdlog::warn("QuantLib vega calculation failed: {}", e.what());
    return 0.0;
  }
}

double OptionChainBuilder::standard_normal_cdf(double x) {
  return 0.5 * std::erfc(-x / std::sqrt(2.0));
}

double OptionChainBuilder::standard_normal_pdf(double x) {
  return std::exp(-0.5 * x * x) / std::sqrt(2.0 * M_PI);
}

std::pair<double, double>
OptionChainBuilder::calculate_d1_d2(double underlying_price, double strike,
                                    double time_to_expiry, double volatility,
                                    double risk_free_rate) {

  double d1 =
      (std::log(underlying_price / strike) +
       (risk_free_rate + 0.5 * volatility * volatility) * time_to_expiry) /
      (volatility * std::sqrt(time_to_expiry));

  double d2 = d1 - volatility * std::sqrt(time_to_expiry);

  return {d1, d2};
}

} // namespace option_chain

File: agents/backend/crates/quant/src/option_chain.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct OptionContract {
    pub symbol: String,
    pub expiry: String,
    pub strike: f64,
    pub is_call: bool,
}

impl OptionContract {
    pub fn new(symbol: &str, expiry: &str, strike: f64, is_call: bool) -> Self {
        Self {
            symbol: symbol.to_string(),
            expiry: expiry.to_string(),
            strike,
            is_call,
        }
    }

    pub fn is_valid(&self) -> bool {
        !self.symbol.is_empty() && self.expiry.len() == 8 && self.strike > 0.0
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct MarketData {
    pub bid: f64,
    pub ask: f64,
    pub last: f64,
    pub volume: i64,
    pub open_interest: i64,
    pub implied_volatility: Option<f64>,
}

impl MarketData {
    pub fn new(bid: f64, ask: f64) -> Self {
        Self {
            bid,
            ask,
            last: (bid + ask) / 2.0,
            volume: 0,
            open_interest: 0,
            implied_volatility: None,
        }
    }

    pub fn mid(&self) -> f64 {
        (self.bid + self.ask) / 2.0
    }

    pub fn spread(&self) -> f64 {
        self.ask - self.bid
    }
}

#[derive(Clone, Debug)]
pub struct OptionChainEntry {
    pub contract: OptionContract,
    pub market_data: MarketData,
    pub theoretical_price: f64,
    pub intrinsic_value: f64,
    pub extrinsic_value: f64,
    pub moneyness: f64,
}

impl OptionChainEntry {
    pub fn new(contract: OptionContract, market_data: MarketData, underlying_price: f64) -> Self {
        let moneyness = if underlying_price > 0.0 {
            contract.strike / underlying_price
        } else {
            1.0
        };

        let intrinsic_value = if contract.is_call {
            (underlying_price - contract.strike).max(0.0)
        } else {
            (contract.strike - underlying_price).max(0.0)
        };

        let mid = market_data.mid();
        let extrinsic_value = (mid - intrinsic_value).max(0.0);

        Self {
            contract,
            market_data,
            theoretical_price: mid,
            intrinsic_value,
            extrinsic_value,
            moneyness,
        }
    }

    pub fn is_valid(&self) -> bool {
        self.contract.is_valid() && self.market_data.bid > 0.0 && self.market_data.ask > 0.0
    }
}

#[derive(Clone, Debug)]
pub struct StrikeChain {
    pub strike: f64,
    pub call: Option<OptionChainEntry>,
    pub put: Option<OptionChainEntry>,
}

impl StrikeChain {
    pub fn new(strike: f64) -> Self {
        Self {
            strike,
            call: None,
            put: None,
        }
    }

    pub fn has_both(&self) -> bool {
        self.call.is_some() && self.put.is_some()
    }

    pub fn call_iv(&self) -> f64 {
        self.call
            .as_ref()
            .and_then(|c| c.market_data.implied_volatility)
            .unwrap_or(0.0)
    }

    pub fn put_iv(&self) -> f64 {
        self.put
            .as_ref()
            .and_then(|p| p.market_data.implied_volatility)
            .unwrap_or(0.0)
    }

    pub fn iv_skew(&self) -> f64 {
        self.put_iv() - self.call_iv()
    }
}

#[derive(Clone, Debug)]
pub struct ExpiryChain {
    symbol: String,
    expiry: String,
    strikes: BTreeMap<i64, StrikeChain>,
}

impl ExpiryChain {
    pub fn new(symbol: &str, expiry: &str) -> Self {
        Self {
            symbol: symbol.to_string(),
            expiry: expiry.to_string(),
            strikes: BTreeMap::new(),
        }
    }

    pub fn add_option(&mut self, entry: OptionChainEntry) {
        let strike_key = (entry.contract.strike * 1000.0) as i64;
        let chain = self
            .strikes
            .entry(strike_key)
            .or_insert_with(|| StrikeChain::new(entry.contract.strike));

        if entry.contract.is_call {
            chain.call = Some(entry);
        } else {
            chain.put = Some(entry);
        }
    }

    pub fn symbol(&self) -> &str {
        &self.symbol
    }

    pub fn expiry(&self) -> &str {
        &self.expiry
    }

    pub fn days_to_expiry(&self) -> i32 {
        let (y, m, d) = match common::expiry::parse_expiry_yyyy_mm_dd(&self.expiry) {
            Ok(ymd) => ymd,
            Err(_) => return 0,
        };
        let year = y as i32;
        let month = time::Month::try_from(m).unwrap_or(time::Month::January);
        let expiry_date = match time::Date::from_calendar_date(year, month, d) {
            Ok(exp) => exp,
            Err(_) => return 0,
        };
        let today = time::OffsetDateTime::now_utc().date();
        (expiry_date - today).whole_days() as i32
    }

    pub fn get_strikes(&self) -> Vec<f64> {
        self.strikes.values().map(|s| s.strike).collect()
    }

    pub fn get_strike_chain(&self, strike: f64) -> Option<&StrikeChain> {
        let strike_key = (strike * 1000.0) as i64;
        self.strikes.get(&strike_key)
    }

    pub fn get_option(&self, strike: f64, is_call: bool) -> Option<&OptionChainEntry> {
        let strike_key = (strike * 1000.0) as i64;
        self.strikes.get(&strike_key).and_then(|chain| {
            if is_call {
                chain.call.as_ref()
            } else {
                chain.put.as_ref()
            }
        })
    }

    pub fn get_strikes_in_range(&self, min_strike: f64, max_strike: f64) -> Vec<f64> {
        self.get_strikes()
            .into_iter()
            .filter(|&s| s >= min_strike && s <= max_strike)
            .collect()
    }

    pub fn find_atm_strike(&self, underlying_price: f64) -> Option<f64> {
        self.get_strikes().into_iter().min_by(|a, b| {
            let dist_a = (a - underlying_price).abs();
            let dist_b = (b - underlying_price).abs();
            dist_a.partial_cmp(&dist_b).unwrap()
        })
    }

    pub fn get_all_options(&self) -> Vec<&OptionChainEntry> {
        self.strikes
            .values()
            .flat_map(|chain| chain.call.as_ref().into_iter().chain(chain.put.as_ref()))
            .collect()
    }

    pub fn get_calls(&self) -> Vec<&OptionChainEntry> {
        self.strikes
            .values()
            .filter_map(|chain| chain.call.as_ref())
            .collect()
    }

    pub fn get_puts(&self) -> Vec<&OptionChainEntry> {
        self.strikes
            .values()
            .filter_map(|chain| chain.put.as_ref())
            .collect()
    }
}

#[derive(Clone, Debug)]
pub struct OptionChain {
    symbol: String,
    underlying_price: f64,
    expiries: BTreeMap<String, ExpiryChain>,
}

impl OptionChain {
    pub fn new(symbol: &str) -> Self {
        Self {
            symbol: symbol.to_string(),
            underlying_price: 0.0,
            expiries: BTreeMap::new(),
        }
    }

    pub fn add_option(&mut self, entry: OptionChainEntry) {
        let expiry = entry.contract.expiry.clone();
        let chain = self
            .expiries
            .entry(expiry)
            .or_insert_with(|| ExpiryChain::new(&self.symbol, &entry.contract.expiry));
        chain.add_option(entry);
    }

    pub fn symbol(&self) -> &str {
        &self.symbol
    }

    pub fn get_expiries(&self) -> Vec<String> {
        self.expiries.keys().cloned().collect()
    }

    pub fn get_expiry_chain(&self, expiry: &str) -> Option<&ExpiryChain> {
        self.expiries.get(expiry)
    }

    pub fn get_expiries_in_dte_range(&self, min_dte: i32, max_dte: i32) -> Vec<String> {
        self.expiries
            .iter()
            .filter(|(_, chain)| {
                let dte = chain.days_to_expiry();
                dte >= min_dte && dte <= max_dte
            })
            .map(|(expiry, _)| expiry.clone())
            .collect()
    }

    pub fn get_all_options(&self) -> Vec<&OptionChainEntry> {
        self.expiries
            .values()
            .flat_map(|chain| chain.get_all_options())
            .collect()
    }

    pub fn set_underlying_price(&mut self, price: f64) {
        self.underlying_price = price;
    }

    pub fn underlying_price(&self) -> f64 {
        self.underlying_price
    }

    pub fn total_option_count(&self) -> usize {
        self.expiries
            .values()
            .map(|c| c.get_all_options().len())
            .sum()
    }

    pub fn expiry_count(&self) -> usize {
        self.expiries.len()
    }
}
